src/backend/tools/files.py

```diff
@@ -0,0 +1,146 @@
+from typing import Any, Dict, List
+
+from pypdf import PdfReader
+
+import backend.crud.file as file_crud
+from backend.chat.collate import combine_documents
+from backend.tools.base import BaseTool
+
+
+class ReadFileTool(BaseTool):
+    """
+    This class reads a file from the file system.
+    """
+
+    MAX_NUM_CHUNKS = 10
+
+    def __init__(self):
+        pass
+
+    @classmethod
+    def is_available(cls) -> bool:
+        return True
+
+    def call(self, parameters: dict, **kwargs: Any) -> List[Dict[str, Any]]:
+        file_name = parameters.get("filename", "")
+        session = kwargs.get("session")
+        user_id = kwargs.get("user_id")
+
+        if not file_name:
+            return []
+
+        files = file_crud.get_files_by_file_names(session, [file_name], user_id)
+
+        if not files:
+            return []
+
+        file = files[0]
+        chunks = chunk_document(file.file_content)
+        result = [
+            {
+                "text": chunk,
+                "title": file.file_name,
+                "url": file.file_path,
+            }
+            for chunk in chunks
+        ]
+
+        return result
+
+
+class SearchFileTool(BaseTool):
+    """
+    This class searches for a query in a file.
+    """
+
+    MAX_NUM_CHUNKS = 10
+
+    def __init__(self):
+        pass
+
+    @classmethod
+    def is_available(cls) -> bool:
+        return True
+
+    def call(self, parameters: dict, **kwargs: Any) -> List[Dict[str, Any]]:
+        query = parameters.get("search_query")
+        file_names = parameters.get("filenames")
+        model_deployment = kwargs.get("model_deployment")
+        session = kwargs.get("session")
+        user_id = kwargs.get("user_id")
+
+        if not query or not file_names:
+            return []
+
+        files = file_crud.get_files_by_file_names(session, file_names, user_id)
+
+        if not files:
+            return []
+
+        files_dicts = []
+        for file in files:
+            chunks = chunk_document(file.file_content)
+            for chunk in chunks:
+                files_dicts.append(
+                    {
+                        "text": chunk,
+                        "title": file.file_name,
+                        "url": file.file_path,
+                    }
+                )
+
+        # Combine and rerank the documents
+        result = combine_documents({query: files_dicts}, model_deployment)
```

Comment:
Like each tool should return only one result and we should chunk, combine and rerank once we have documents for every tool?

Reply:
Yes, but each tool returns multiple outputs/documents and we should chunk, combine and rerank them

Reply:
Since we're moving away from using documents and are now using tool_restuls, what should we use to rerank the outputs? Some tools might not have a query 🤔

Reply:
Oh yes, we should only rerank on internet search / ones with a query - but chunk on them all to my knowledge